From 7592330ca5665e732df79dc44ffbf091baf141d7 Mon Sep 17 00:00:00 2001
From: Sora Su <sora_su@wiwynn.com>
Date: Fri, 13 Mar 2026 12:58:37 +0800
Subject: [PATCH 01/19] oem-meta: Add file IO responder for event logs from BIC

Add handler for write-file command "EVENT_LOG", which is sent when host
events are triggered, parse event type from a list and add SEL in BMC.

Tested: Verified on the Yosemite4 platform. All event were recorded
correctly.

Change-Id: I86d90cea9bbf61c5415d1c85e17a23d833d815f2
Signed-off-by: Bright Cheng <bright_cheng@wiwynn.com>
Signed-off-by: Liora Guo <liora.guo.wiwynn@gmail.com>
Signed-off-by: Rush Chen <Rush_Chen@wiwynn.com>
Signed-off-by: Unive Tien <unive.tien.wiwynn@gmail.com>
Signed-off-by: Zoey YJ Chung <zoey.yj.chung.wiwynn@gmail.com>
Signed-off-by: Kay YC Huang <kay.yc.huang.wiwynn@gmail.com>
Signed-off-by: Sora Su <sora_su@wiwynn.com>
Signed-off-by: Patrick Williams <patrick@stwcx.xyz>
---
 oem/meta/libpldmresponder/file_io.cpp         |   3 +
 oem/meta/libpldmresponder/file_io_by_type.hpp |   3 +-
 .../file_io_type_event_log.cpp                | 719 ++++++++++++++++++
 .../file_io_type_event_log.hpp                |  39 +
 oem/meta/libpldmresponder/meson.build         |   1 +
 5 files changed, 764 insertions(+), 1 deletion(-)
 create mode 100644 oem/meta/libpldmresponder/file_io_type_event_log.cpp
 create mode 100644 oem/meta/libpldmresponder/file_io_type_event_log.hpp

diff --git a/oem/meta/libpldmresponder/file_io.cpp b/oem/meta/libpldmresponder/file_io.cpp
index 2293d43..8949674 100644
--- a/oem/meta/libpldmresponder/file_io.cpp
+++ b/oem/meta/libpldmresponder/file_io.cpp
@@ -1,6 +1,7 @@
 #include "file_io.hpp"
 
 #include "file_io_type_bios_version.hpp"
+#include "file_io_type_event_log.hpp"
 #include "file_io_type_http_boot.hpp"
 #include "file_io_type_post_code.hpp"
 #include "file_io_type_power_control.hpp"
@@ -26,6 +27,8 @@ std::unique_ptr<FileHandler> FileIOHandler::getHandlerByType(
                                                          dBusHandler);
         case FileIOType::HTTP_BOOT:
             return std::make_unique<HttpBootHandler>();
+        case FileIOType::EVENT_LOG:
+            return std::make_unique<EventLogHandler>(messageTid);
         default:
             error("Get invalid file io type {FILEIOTYPE}", "FILEIOTYPE",
                   fileIOType);
diff --git a/oem/meta/libpldmresponder/file_io_by_type.hpp b/oem/meta/libpldmresponder/file_io_by_type.hpp
index 86cb708..6f0ae39 100644
--- a/oem/meta/libpldmresponder/file_io_by_type.hpp
+++ b/oem/meta/libpldmresponder/file_io_by_type.hpp
@@ -16,7 +16,8 @@ enum class FileIOType : uint8_t
     BIOS_VERSION = 0x01,
     POWER_CONTROL = 0x02,
     HTTP_BOOT = 0x03,
-    APML_ALERT = 0x04
+    APML_ALERT = 0x04,
+    EVENT_LOG = 0x05
 };
 
 /**
diff --git a/oem/meta/libpldmresponder/file_io_type_event_log.cpp b/oem/meta/libpldmresponder/file_io_type_event_log.cpp
new file mode 100644
index 0000000..d0ebe95
--- /dev/null
+++ b/oem/meta/libpldmresponder/file_io_type_event_log.cpp
@@ -0,0 +1,719 @@
+#include "file_io_type_event_log.hpp"
+
+#include <phosphor-logging/commit.hpp>
+#include <phosphor-logging/lg2.hpp>
+#include <sdbusplus/bus.hpp>
+#include <xyz/openbmc_project/State/Power/event.hpp>
+#include <xyz/openbmc_project/State/Thermal/event.hpp>
+
+#include <array>
+#include <format>
+#include <functional>
+#include <map>
+#include <string>
+#include <string_view>
+#include <utility>
+
+PHOSPHOR_LOG2_USING;
+
+namespace pldm::responder::oem_meta
+{
+static constexpr size_t EVENT_LEN = 5;
+static constexpr size_t EVENT_WITH_TIMESTAMP_LEN = 9;
+
+enum class EventType
+{
+    CPU_THERMAL_TRIP = 0x0,
+    HSC_OCP = 0x1,
+    P12V_STBY_UV = 0x2,
+    PMALERT_ASSERTION = 0x3,
+    FAST_PROCHOT_ASSERTION = 0x4,
+    FRB3_TIMER_EXPIRED = 0x5,
+    POWER_ON_SEQUENCE_FAILURE = 0x6,
+    DIMM_PMIC_ERROR = 0x7,
+    ADDC_DUMP = 0x8,
+    BMC_COMES_OUT_OF_COLD_RESET = 0x9,
+    BIOS_FRB12_WATCHDOG_TIMER_EXPIRED = 0xa,
+    BIC_POWER_FAILURE = 0xb,
+    CPU_POWER_FAILURE = 0xc,
+    V_BOOT_FAILURE = 0xd,
+    BMC_IS_REQUESTED_TO_REBOOT = 0xe,
+    CHASSIS_POWER_ON_IS_INITIATED_BY_NIC_INSERTION = 0xf,
+    BLADE_IS_POWERED_CYCLE_BY_BLADE_BUTTON = 0x10,
+    CHASSIS_IS_POWERED_CYCLE_BY_SLED_BUTTON = 0x11,
+    HSC_FAULT = 0x12,
+    SYS_THROTTLE = 0x13,
+    VR_FAULT = 0x14,
+    SYSTEM_MANAGEMENT_ERROR = 0x15,
+    POST_COMPLETED = 0x16,
+    FAN_ERROR = 0x17,
+    HDT_PRSNT_ASSERTION = 0x18,
+    PLTRST_ASSERTION = 0x19,
+    APML_ALERT_ASSERTION = 0x1a,
+    CXL1_HB = 0x1b,
+    CXL2_HB = 0x1c,
+    POST_STARTED = 0x1d,
+    POST_ENDED = 0x1e,
+    PROCHOT_IS_TRIGGERED_DUE_TO_SD_SENSOR_READING_EXCEED_UCR = 0x1f,
+    FRB2_WDT_HARD_RST = 0x20,
+    FRB2_WDT_PWR_DOWN = 0x21,
+    FRB2_WDT_PWR_CYCLE = 0x22,
+    OS_LOAD_WDT_EXPIRED = 0x23,
+    OS_LOAD_WDT_HARD_RST = 0x24,
+    OS_LOAD_WDT_PWR_DOWN = 0x25,
+    OS_LOAD_WDT_PWR_CYCLE = 0x26,
+    MTIA_FAULT = 0x27,
+    POST_TIMEOUTED = 0x28,
+};
+
+enum class EventAssert : uint8_t
+{
+    DEASSERTED = 0x0,
+    ASSERTED = 0x1,
+};
+
+struct VRSource
+{
+    std::string netName;
+    bool isPMBus;
+};
+
+static constexpr auto eventList = std::to_array({
+    "CPU Thermal Trip",
+    "HSC OCP",
+    "P12V_STBY UV",
+    "PMALERT Assertion",
+    "FAST_PROCHOT Assertion",
+    "FRB3 Timer Expired",
+    "Power-On Sequence Failure",
+    "DIMM PMIC Error",
+    "ADDC dump",
+    "BMC comes out of cold reset",
+    "BIOS FRB2 Watchdog Timer Expired",
+    "BIC Power Failure",
+    "CPU Power Failure",
+    "v-boot Failure",
+    "BMC is requested to reboot",
+    "Chassis power on is initiated by NIC insertion",
+    "Blade is powered cycle by blade button",
+    "Chassis is powered cycle by sled button",
+    "HSC Fault",
+    "SYS Throttle",
+    "VR Fault",
+    "System Management Error",
+    "Post-Completed",
+    "Fan Error",
+    "HDT_PRSNT Assertion",
+    "PLTRST Assertion",
+    "APML_ALERT Assertion",
+    "CXL1 HB",
+    "CXL2 HB",
+    "Post-Started",
+    "Post-Ended",
+    "PROCHOT is triggered due to SD sensor reading exceed UCR",
+    "FRB2 WDT HARD RST",
+    "FRB2 WDT PWR DOWN",
+    "FRB2 WDT PWR CYCLE",
+    "OS LOAD WDT EXPIRED",
+    "OS LOAD WDT HARD_RST",
+    "OS LOAD WDT PWR_DOWN",
+    "OS LOAD WDT PWR_CYCLE",
+    "MTIA FAULT",
+    "Post-Timeouted",
+});
+
+static constexpr auto pmic_event_type = std::to_array({
+    "SWAout OV",
+    "SWBout OV",
+    "SWCout OV",
+    "SWDout OV",
+    "Vin Bulk OV",
+    "Vin Mgmt OV",
+    "SWAout UV",
+    "SWBout UV",
+    "SWCout UV",
+    "SWDout UV",
+    "Vin Bulk UV",
+    "Vin Mgmt To Vin Buck Switchover",
+    "High Temp Warning",
+    "Vout 1v8 PG",
+    "High Current Warning",
+    "Current Limit Warning",
+    "Critical Temp Shutdown",
+});
+
+static constexpr auto dimm_id = std::to_array({
+    "DIMM A0",
+    "DIMM A1",
+    "DIMM A2",
+    "DIMM A3",
+    "DIMM A4",
+    "DIMM A5",
+    "DIMM A6",
+    "DIMM A7",
+    "DIMM A8",
+    "DIMM A9",
+    "DIMM A10",
+    "DIMM A11",
+});
+
+static const auto vr_source = std::to_array<VRSource>({
+    {"PVDDCR_CPU0", true},    {"PVDDCR_SOC", true},
+    {"PVDDCR_CPU1", true},    {"PVDDIO", true},
+    {"PVDD11_S3", true},      {"PVDDQ_AB_ASIC1", true},
+    {"P0V85_ASIC1", true},    {"P0V8_ASIC1", true},
+    {"PVDDQ_CD_ASIC1", true}, {"PVDDQ_AB_ASIC2", true},
+    {"P0V85_ASIC2", true},    {"PVDDQ_CD_ASIC2", true},
+    {"P0V8_ASIC2", true},     {"P1V5_RETIMER_1", false},
+    {"P0V9_STBY_1", false},   {"P3V3_E1S_0", false},
+    {"P3V3_E1S_1", false},    {"P12V_E1S_0", false},
+    {"P12V_E1S_1", false},    {"PVTT_AB_ASIC1", false},
+    {"PVTT_AB_ASIC2", false}, {"PVTT_CD_ASIC1", false},
+    {"PVTT_CD_ASIC2", false}, {"PVPP_AB_ASIC1", false},
+    {"PVPP_AB_ASIC2", false}, {"PVPP_CD_ASIC1", false},
+    {"PVPP_CD_ASIC2", false},
+});
+
+struct FaultDesc
+{
+    std::string name;
+    std::string object_path;
+    const char* type_label;
+    using RawRecorder = void (*)(const std::string& object_path,
+                                 EventAssert eventStatus,
+                                 const std::string& failureData);
+    RawRecorder recorder;
+};
+
+namespace record
+{
+template <const char* TypeLabel, typename AssertType, typename DeassertType>
+static void commit(const std::string& object_path, EventAssert eventStatus,
+                   const std::string& failureData)
+{
+    if (eventStatus == EventAssert::ASSERTED)
+    {
+        lg2::commit(
+            AssertType(TypeLabel, object_path, "FAILURE_DATA", failureData));
+    }
+    else
+    {
+        lg2::commit(DeassertType(TypeLabel, object_path));
+    }
+}
+
+static void powerRailFault(const std::string& object_path,
+                           EventAssert eventStatus,
+                           const std::string& failureData)
+{
+    static constexpr char label[] = "POWER_RAIL";
+
+    using Fault =
+        sdbusplus::error::xyz::openbmc_project::state::Power::PowerRailFault;
+    using Normal = sdbusplus::event::xyz::openbmc_project::state::Power::
+        PowerRailFaultRecovered;
+
+    commit<label, Fault, Normal>(object_path, eventStatus, failureData);
+}
+
+static void voltageRegulatorFault(const std::string& object_path,
+                                  EventAssert eventStatus,
+                                  const std::string& failureData)
+{
+    static constexpr char label[] = "VOLTAGE_REGULATOR";
+
+    using Fault = sdbusplus::error::xyz::openbmc_project::state::Power::
+        VoltageRegulatorFault;
+    using Normal = sdbusplus::event::xyz::openbmc_project::state::Power::
+        VoltageRegulatorFaultRecovered;
+
+    commit<label, Fault, Normal>(object_path, eventStatus, failureData);
+}
+
+static void thermalFault(const std::string& object_path,
+                         EventAssert eventStatus,
+                         const std::string& failureData)
+{
+    static constexpr char label[] = "DEVICE";
+
+    using Fault = sdbusplus::error::xyz::openbmc_project::state::Thermal::
+        DeviceOverOperatingTemperatureFault;
+    using Normal = sdbusplus::event::xyz::openbmc_project::state::Thermal::
+        DeviceOperatingNormalTemperature;
+
+    commit<label, Fault, Normal>(object_path, eventStatus, failureData);
+}
+} // namespace record
+
+static const std::map<uint8_t, FaultDesc> mtiaFaultsList = {
+    {0x00,
+     {"MTIA_P3V3",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p3v3",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x01,
+     {"MTIA_P0V85_PVDD",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v85_pvdd",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x02,
+     {"MTIA_P0V75_PVDD_CH_N",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v75_pvdd_ch_n",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x03,
+     {"MTIA_P0V75_MAX_PHY_N",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v75_max_phy_n",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x04,
+     {"MTIA_P0V75_PVDD_CH_S",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v75_pvdd_ch_s",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x05,
+     {"MTIA_P0V75_MAX_PHY_S",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v75_max_phy_s",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x06,
+     {"MTIA_P0V75_TRVDD_ZONEA",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v75_trvdd_zonea",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x07,
+     {"MTIA_P1V8_VPP_HBM0_HBM2_HBM4",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p1v8_vpp_hbm0_hbm2_hbm4",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x08,
+     {"MTIA_P0V75_TRVDD_ZONEB",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v75_trvdd_zoneb",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x09,
+     {"MTIA_P0V4_VDDQL_HBM0_HBM2_HBM4",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v4_vddql_hbm0_hbm2_hbm4",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x0A,
+     {"MTIA_P1V1_VDDC_HBM0_HBM2_HBM4",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p1v1_vddc_hbm0_hbm2_hbm4",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x0B,
+     {"MTIA_P0V75_VDDPHY_HBM0_HBM2_HBM4",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v75_vddphy_hbm0_hbm2_hbm4",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x0C,
+     {"MTIA_P0V9_TRVDD_ZONEA",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v9_trvdd_zonea",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x0D,
+     {"MTIA_P1V8_VPP_HBM1_HBM3_HBM5",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p1v8_vpp_hbm1_hbm3_hbm5",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x0E,
+     {"MTIA_P0V9_TRVDD_ZONEB",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v9_trvdd_zoneb",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x0F,
+     {"MTIA_P0V4_VDDQL_HBM1_HBM3_HBM5",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v4_vddql_hbm1_hbm3_hbm5",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x10,
+     {"MTIA_P1V1_VDDC_HBM1_HBM3_HBM5",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p1v1_vddc_hbm1_hbm3_hbm5",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x11,
+     {"MTIA_P0V75_VDDPHY_HBM1_HBM3_HBM5",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v75_vddphy_hbm1_hbm3_hbm5",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x12,
+     {"MTIA_P0V8_VDDA_PCIE",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v8_vdda_pcie",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x13,
+     {"MTIA_P1V2_VDDHTX_PCIE",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p1v2_vddhtx_pcie",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x14,
+     {"MTIA_P12V_UBC1",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p12v_ubc1",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x15,
+     {"MTIA_P12V_UBC2",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p12v_ubc2",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x16,
+     {"MTIA_VR_MAX",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/vr_max",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x30,
+     {"MTIA_PLL_VDDA15_PCIE_MAX_CORE",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/pll_vdda15_pcie_max_core",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x31,
+     {"MTIA_P0V75_AVDD_HSCL",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v75_avdd_hscl",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x32,
+     {"MTIA_P0V75_VDDC_CLKOBS",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p0v75_vddc_clkobs",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x33,
+     {"MTIA_PLL_VDDA15_MAX_CORE_N",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/pll_vdda15_max_core_n",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x34,
+     {"MTIA_PVDD1P5_S",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/pvdd1p5_s",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x35,
+     {"MTIA_PVDD1P5_N",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/pvdd1p5_n",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x36,
+     {"MTIA_PVDD0P9_S",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/pvdd0p9_s",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x37,
+     {"MTIA_PVDD0P9_N",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/pvdd0p9_n",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x38,
+     {"MTIA_PLL_VDDA15_HBM0_HBM2_HBM4",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/pll_vdda15_hbm0_hbm2_hbm4",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x39,
+     {"MTIA_PLL_VDDA15_HBM1_HBM3_HBM5",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/pll_vdda15_hbm1_hbm3_hbm5",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x3A,
+     {"MTIA_P3V3_OSC",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p3v3_osc",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x3B,
+     {"MTIA_P5V",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/p5v",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x3C,
+     {"MTIA_LDO_IN_1V8",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/ldo_in_1v8",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x3D,
+     {"MTIA_LDO_IN_1V2",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/ldo_in_1v2",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x3E,
+     {"MTIA_PLL_VDDA15_MAX_CORE_S",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/pll_vdda15_max_core_s",
+      "VOLTAGE_REGULATOR", record::voltageRegulatorFault}},
+    {0x50,
+     {"MTIA_POWER_ON_SEQUENCE_FAIL",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/asic0",
+      "POWER_RAIL", record::powerRailFault}},
+    {0x51,
+     {"MTIA_FM_ASIC_0_THERMTRIP_N",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/asic_0_thermtrip_n",
+      "DEVICE", record::thermalFault}},
+    {0x52,
+     {"MTIA_FM_ATH_PLD_HBM3_CATTRIP",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/ath_pld_hbm3_cattrip",
+      "DEVICE", record::thermalFault}},
+    {0x53,
+     {"MTIA_VR_FAULT_CAUSE_POWER_DOWN",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/vr_fault_power_down",
+      "POWER_RAIL", record::powerRailFault}},
+    {0x54,
+     {"MTIA_ATH_GPIO_3",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/ath_gpio_3",
+      "DEVICE", record::thermalFault}},
+    {0x55,
+     {"MTIA_ATH_GPIO_4",
+      "/xyz/openbmc_project/inventory/system/board/Minerva_Aegis/ath_gpio_4",
+      "DEVICE", record::thermalFault}},
+};
+
+namespace report
+{
+static void recordEventLog(const FaultDesc& fault, EventAssert eventStatus,
+                           uint8_t detail1, uint8_t detail2)
+{
+    if (!fault.recorder)
+    {
+        return;
+    }
+
+    auto failureData =
+        std::format("{} - [{:02X} {:02X}]", fault.name, detail1, detail2);
+
+    fault.recorder(fault.object_path, eventStatus, failureData);
+}
+
+static void thermalFault(const message& eventData, const std::string&,
+                         const std::string& slotNum)
+{
+    auto device = "/xyz/openbmc_project/State/Thermal/host" + slotNum + "/cpu0";
+    EventAssert eventStatus = static_cast<EventAssert>(eventData[1]);
+
+    record::thermalFault(device, eventStatus, "Thermal Trip");
+}
+
+static void powerRailFault(const message& eventData, const std::string&,
+                           const std::string& slotNum)
+{
+    auto device = "/xyz/openbmc_project/State/Power/host" + slotNum + "/power";
+    EventAssert eventStatus = static_cast<EventAssert>(eventData[1]);
+
+    record::powerRailFault(device, eventStatus, "Power Rail Fault");
+}
+
+static void prochotSdSensor(const message& eventData, const std::string&,
+                            const std::string& slotNum)
+{
+    auto device = "/xyz/openbmc_project/State/Thermal/host" + slotNum + "/cpu0";
+
+    if (eventData[3] == 0 && eventData[4] == 0)
+    {
+        using sdbusplus::event::xyz::openbmc_project::state::Thermal::
+            DeviceOperatingNormalTemperature;
+
+        lg2::commit(DeviceOperatingNormalTemperature("DEVICE", device));
+    }
+    else
+    {
+        using sdbusplus::error::xyz::openbmc_project::state::Thermal::
+            DeviceOverOperatingTemperatureFault;
+
+        auto failureData =
+            std::format("SD sensor reading exceed UCR, data: 0x{:02x}{:02x}",
+                        eventData[3], eventData[4]);
+
+        lg2::commit(DeviceOverOperatingTemperatureFault(
+            "DEVICE", device, "FAILURE_DATA", failureData));
+    }
+}
+
+static void mtiaFault(const message& eventData, const std::string&,
+                      const std::string&)
+{
+    EventAssert eventStatus = static_cast<EventAssert>(eventData[1]);
+    auto faultEvent = mtiaFaultsList.find(eventData[2]);
+    if (faultEvent == std::end(mtiaFaultsList))
+    {
+        lg2::error("Invalid MTIA fault source 0x{SOURCE}", "SOURCE", lg2::hex,
+                   eventData[2]);
+        return;
+    }
+
+    recordEventLog(faultEvent->second, eventStatus, eventData[3], eventData[4]);
+}
+
+static void reportError(const message&, const std::string& event,
+                        const std::string&)
+{
+    pldm::utils::reportError(event.c_str());
+}
+} // namespace report
+
+namespace format
+{
+static std::string vrFault(const message& eventData, EventType eventType)
+{
+    if (eventData[2] >= vr_source.size())
+    {
+        lg2::error("Invalid vr source {VR_SOURCE}", "VR_SOURCE", eventData[2]);
+        return {};
+    }
+
+    if (vr_source[eventData[2]].isPMBus)
+    {
+        return std::format("{} : {} status: 0x{:02x}{:02x}",
+                           eventList[std::to_underlying(eventType)],
+                           vr_source[eventData[2]].netName, eventData[3],
+                           eventData[4]);
+    }
+    return std::format("{} : {}", eventList[std::to_underlying(eventType)],
+                       vr_source[eventData[2]].netName);
+}
+
+static std::string dimmPmicError(const message& eventData, EventType eventType)
+{
+    if ((eventData[2] >= dimm_id.size()) ||
+        (eventData[3] >= pmic_event_type.size()))
+    {
+        lg2::error(" Invalid dimm id {ID}, pmic event type {TYPE}", "ID",
+                   eventData[2], "TYPE", eventData[3]);
+        return {};
+    }
+
+    return eventList[std::to_underlying(eventType)] +
+           std::format(" : {} {}", dimm_id[eventData[2]],
+                       pmic_event_type[eventData[3]]);
+}
+} // namespace format
+
+using EventReporter = void (*)(const message&, const std::string&,
+                               const std::string&);
+using EventFormatter = std::string (*)(const message&, EventType);
+
+struct EventDescriptor
+{
+    EventReporter reporter;
+    EventFormatter formatter;
+};
+
+static const std::map<EventType, EventDescriptor> eventHandlers = {
+    {EventType::CPU_THERMAL_TRIP, {report::thermalFault, nullptr}},
+    {EventType::FAST_PROCHOT_ASSERTION, {report::thermalFault, nullptr}},
+    {EventType::HSC_OCP, {report::powerRailFault, nullptr}},
+    {EventType::P12V_STBY_UV, {report::powerRailFault, nullptr}},
+    {EventType::POWER_ON_SEQUENCE_FAILURE, {report::powerRailFault, nullptr}},
+    {EventType::DIMM_PMIC_ERROR,
+     {report::powerRailFault, format::dimmPmicError}},
+    {EventType::CXL1_HB, {nullptr, nullptr}},
+    {EventType::CXL2_HB, {nullptr, nullptr}},
+    {EventType::PLTRST_ASSERTION, {nullptr, nullptr}},
+    {EventType::POST_STARTED, {nullptr, nullptr}},
+    {EventType::POST_ENDED, {nullptr, nullptr}},
+    {EventType::PROCHOT_IS_TRIGGERED_DUE_TO_SD_SENSOR_READING_EXCEED_UCR,
+     {report::prochotSdSensor, nullptr}},
+    {EventType::MTIA_FAULT, {report::mtiaFault, nullptr}},
+    {EventType::PMALERT_ASSERTION, {report::reportError, format::vrFault}},
+    {EventType::FRB3_TIMER_EXPIRED, {report::reportError, nullptr}},
+    {EventType::ADDC_DUMP, {report::reportError, nullptr}},
+    {EventType::BMC_COMES_OUT_OF_COLD_RESET, {report::reportError, nullptr}},
+    {EventType::BIOS_FRB12_WATCHDOG_TIMER_EXPIRED,
+     {report::reportError, nullptr}},
+    {EventType::BIC_POWER_FAILURE, {report::reportError, nullptr}},
+    {EventType::CPU_POWER_FAILURE, {report::reportError, nullptr}},
+    {EventType::V_BOOT_FAILURE, {report::reportError, nullptr}},
+    {EventType::BMC_IS_REQUESTED_TO_REBOOT, {report::reportError, nullptr}},
+    {EventType::CHASSIS_POWER_ON_IS_INITIATED_BY_NIC_INSERTION,
+     {report::reportError, nullptr}},
+    {EventType::BLADE_IS_POWERED_CYCLE_BY_BLADE_BUTTON,
+     {report::reportError, nullptr}},
+    {EventType::CHASSIS_IS_POWERED_CYCLE_BY_SLED_BUTTON,
+     {report::reportError, nullptr}},
+    {EventType::HSC_FAULT, {report::reportError, nullptr}},
+    {EventType::SYS_THROTTLE, {report::reportError, nullptr}},
+    {EventType::VR_FAULT, {report::reportError, format::vrFault}},
+    {EventType::SYSTEM_MANAGEMENT_ERROR, {report::reportError, nullptr}},
+    {EventType::POST_COMPLETED, {report::reportError, nullptr}},
+    {EventType::FAN_ERROR, {report::reportError, nullptr}},
+    {EventType::HDT_PRSNT_ASSERTION, {report::reportError, nullptr}},
+    {EventType::APML_ALERT_ASSERTION, {report::reportError, nullptr}},
+    {EventType::FRB2_WDT_HARD_RST, {report::reportError, nullptr}},
+    {EventType::FRB2_WDT_PWR_DOWN, {report::reportError, nullptr}},
+    {EventType::FRB2_WDT_PWR_CYCLE, {report::reportError, nullptr}},
+    {EventType::OS_LOAD_WDT_EXPIRED, {report::reportError, nullptr}},
+    {EventType::OS_LOAD_WDT_HARD_RST, {report::reportError, nullptr}},
+    {EventType::OS_LOAD_WDT_PWR_DOWN, {report::reportError, nullptr}},
+    {EventType::OS_LOAD_WDT_PWR_CYCLE, {report::reportError, nullptr}},
+    {EventType::POST_TIMEOUTED, {report::reportError, nullptr}},
+};
+
+static void checkEventAssert(const message& eventData, std::string& event,
+                             const std::string& slotNum)
+{
+    event.insert(0, std::format("Event: Host{} ", slotNum));
+
+    EventAssert eventStatus = static_cast<EventAssert>(eventData[1]);
+    switch (eventStatus)
+    {
+        case EventAssert::ASSERTED:
+            event.append(", ASSERTED");
+            break;
+        case EventAssert::DEASSERTED:
+            event.append(", DEASSERTED");
+            break;
+    }
+}
+
+static std::string processEventData(const message& eventData,
+                                    const std::string& slotNum)
+{
+    std::string event;
+    EventType eventType = static_cast<EventType>(eventData[0]);
+
+    auto it = eventHandlers.find(eventType);
+    if (it != eventHandlers.end() && it->second.formatter != nullptr)
+    {
+        event = it->second.formatter(eventData, eventType);
+    }
+    else
+    {
+        if (static_cast<std::size_t>(std::to_underlying(eventType)) <
+            eventList.size())
+        {
+            event = eventList[std::to_underlying(eventType)];
+        }
+        else
+        {
+            event = "Event type not supported";
+        }
+    }
+
+    checkEventAssert(eventData, event, slotNum);
+
+    return event;
+}
+
+static void addSystemEventLogAndJournal(const message& eventData,
+                                        const std::string& event,
+                                        const std::string& slotNum)
+{
+    EventType eventType = static_cast<EventType>(eventData[0]);
+    auto it = eventHandlers.find(eventType);
+    if (it != eventHandlers.end())
+    {
+        if (it->second.reporter)
+        {
+            it->second.reporter(eventData, event, slotNum);
+        }
+    }
+    else
+    {
+        lg2::error("Unexpected error: {ERROR}", "ERROR", event);
+    }
+}
+
+int EventLogHandler::write(const message& eventData)
+{
+    static std::map<std::pair<pldm_tid_t, EventType>, uint32_t>
+        receivedEventTimeStamp;
+
+    if (eventData.size() != EVENT_LEN &&
+        eventData.size() != EVENT_WITH_TIMESTAMP_LEN)
+    {
+        lg2::error(
+            "Invalid incoming data for event log, data size {SIZE} bytes",
+            "SIZE", eventData.size());
+        return PLDM_ERROR;
+    }
+
+    if (eventData[0] >= eventList.size())
+    {
+        lg2::error("Invalid event type for event log, event type {TYPE}",
+                   "TYPE", eventData[0]);
+        return PLDM_ERROR;
+    }
+
+    if (eventData.size() == EVENT_WITH_TIMESTAMP_LEN)
+    {
+        uint32_t eventTimeStamp =
+            (static_cast<uint32_t>(eventData[5]) << 12) +
+            (static_cast<uint32_t>(eventData[6]) << 8) +
+            (static_cast<uint32_t>(eventData[7]) << 4) +
+            static_cast<uint32_t>(eventData[8]);
+        auto eventKey = std::pair<pldm_tid_t, EventType>(
+            tid, static_cast<EventType>(eventData[0]));
+        if (receivedEventTimeStamp.contains(eventKey) &&
+            receivedEventTimeStamp[eventKey] == eventTimeStamp)
+        {
+            lg2::info(
+                "oem event: TID {TID}: Event time stamp {EVTT} skip retry event",
+                "TID", tid, "EVTT", eventTimeStamp);
+            return PLDM_SUCCESS;
+        }
+        receivedEventTimeStamp[eventKey] = eventTimeStamp;
+    }
+
+    std::string slotNum = pldm::oem_meta::getSlotNumberStringByTID(tid);
+    std::string event = processEventData(eventData, slotNum);
+    addSystemEventLogAndJournal(eventData, event, slotNum);
+
+    return PLDM_SUCCESS;
+}
+
+} // namespace pldm::responder::oem_meta
diff --git a/oem/meta/libpldmresponder/file_io_type_event_log.hpp b/oem/meta/libpldmresponder/file_io_type_event_log.hpp
new file mode 100644
index 0000000..b58061c
--- /dev/null
+++ b/oem/meta/libpldmresponder/file_io_type_event_log.hpp
@@ -0,0 +1,39 @@
+#pragma once
+
+#include "common/utils.hpp"
+#include "file_io_by_type.hpp"
+#include "oem/meta/utils.hpp"
+
+namespace pldm::responder::oem_meta
+{
+
+/** @class EventLogHandler
+ *
+ *  @brief Inherits and implements FileHandler. This class is used
+ *  to parse Event logs from BIC and add SEL
+ */
+class EventLogHandler : public FileHandler
+{
+  public:
+    EventLogHandler() = delete;
+    EventLogHandler(const EventLogHandler&) = delete;
+    EventLogHandler(EventLogHandler&&) = delete;
+    EventLogHandler& operator=(const EventLogHandler&) = delete;
+    EventLogHandler& operator=(EventLogHandler&&) = delete;
+
+    explicit EventLogHandler(pldm_tid_t tid) : tid(tid) {}
+
+    ~EventLogHandler() = default;
+
+    /** @brief Method to parse event log from eventList
+     *  @param[in] data - eventData
+     *  @return  PLDM status code
+     */
+    int write(const message& data);
+
+  private:
+    /** @brief The terminus ID of the message source*/
+    pldm_tid_t tid = 0;
+};
+
+} // namespace pldm::responder::oem_meta
diff --git a/oem/meta/libpldmresponder/meson.build b/oem/meta/libpldmresponder/meson.build
index 265f6f3..62fecbc 100644
--- a/oem/meta/libpldmresponder/meson.build
+++ b/oem/meta/libpldmresponder/meson.build
@@ -1,6 +1,7 @@
 oem_files += files(
     'file_io.cpp',
     'file_io_type_bios_version.cpp',
+    'file_io_type_event_log.cpp',
     'file_io_type_http_boot.cpp',
     'file_io_type_post_code.cpp',
     'file_io_type_power_control.cpp',
-- 
2.52.0

